File: src/main/java/com/back/nbe9112team06/domain/meeting/service/MeetingService.java
package com.back.nbe9112team06.domain.meeting.service;

import com.back.nbe9112team06.domain.meeting.dto.ConfirmedScheduleResponse;
import com.back.nbe9112team06.domain.meeting.dto.FinalizeRequest;
import com.back.nbe9112team06.domain.meeting.dto.request.MeetingCreateRequest;
import com.back.nbe9112team06.domain.meeting.dto.response.MeetingCreateResponse;
import com.back.nbe9112team06.domain.meeting.dto.response.MeetingEntryResponse;
import com.back.nbe9112team06.domain.meeting.entity.Meeting;
import com.back.nbe9112team06.domain.meeting.entity.MeetingStatus;
import com.back.nbe9112team06.domain.meeting.entity.MeetingsDate;
import com.back.nbe9112team06.domain.meeting.repository.MeetingRepository;
import com.back.nbe9112team06.domain.member.entity.Member;
import com.back.nbe9112team06.domain.member.service.MemberService;
import com.back.nbe9112team06.global.error.ErrorCode;
import com.back.nbe9112team06.global.exception.BusinessException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class MeetingService {

    private static final String URL_CHAR_POOL = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    private static final int URL_LENGTH = 10;

    private final MeetingRepository meetingRepository;
    private final MemberService memberService;
    private final SecureRandom secureRandom = new SecureRandom();

    // ── 모임 생성 ──────────────────────────────
    @Transactional
    public MeetingCreateResponse createMeeting(Integer memberId, MeetingCreateRequest request) {

        Member member = memberService.findById(memberId)
                .orElseThrow(() -> new BusinessException(ErrorCode.MEMBER_NOT_FOUND));

        String randomUrl = generateUniqueUrl();
        Meeting meeting = Meeting.create(
                request.title(),
                request.category(),
                request.duration(),
                member,
                randomUrl
        );

        for (LocalDate date : request.dates()) {
            MeetingsDate meetingsDate = MeetingsDate.create(date, member.getEmail());
            meeting.addMeetingsDate(meetingsDate);
        }

        Meeting saved = meetingRepository.save(meeting);
        return new MeetingCreateResponse(saved.getId(), saved.getRandomUrl());
    }

    @Transactional(readOnly = true)
    public MeetingEntryResponse getMeetingByRandomUrl(String randomUrl) {
        Meeting meeting = findMeetingByRandomUrlInternal(randomUrl);

        List<LocalDate> dates = meeting.getMeetingsDates().stream()
                .map(MeetingsDate::getDate)
                .sorted()
                .toList();

        return new MeetingEntryResponse(
                meeting.getId(),
                meeting.getTitle(),
                meeting.getCategory(),
                meeting.getDuration(),
                meeting.getStatus(),
                meeting.getRandomUrl(),
                dates,
                meeting.getCreatedAt(),
                meeting.getConfirmedDate(),
                meeting.getConfirmedTime()
        );
    }

    // ── 모임 삭제 ──────────────────────────────
    @Transactional
    public void deleteMeeting(Integer meetingId, Integer memberId) {
        Meeting meeting = findMeetingInternal(meetingId);

        if (!meeting.isHost(memberId)) {
            throw new BusinessException(ErrorCode.NOT_MEETING_HOST);
        }

        meetingRepository.delete(meeting);
    }

    @Transactional(readOnly = true)
    public boolean checkIsHost(String randomUrl, Integer memberId) {
        Meeting meeting = findMeetingByRandomUrlInternal(randomUrl);
        return meeting.isHost(memberId);
    }

    // ── 일정 확정 ──────────────────────────────
    @Transactional
    public ConfirmedScheduleResponse confirm(Integer meetingId, Integer memberId, FinalizeRequest request) {
        Meeting meeting = findMeetingInternal(meetingId);

        if (!meeting.isHost(memberId)) {
            throw new BusinessException(ErrorCode.NOT_MEETING_HOST);
        }

        if (meeting.getParticipants().isEmpty()) {
            throw new BusinessException(ErrorCode.MEETING_NO_PARTICIPANTS);
        }

        if (meeting.getStatus() == MeetingStatus.CONFIRMED) {
            throw new BusinessException(ErrorCode.ALREADY_CONFIRMED);
        }

        meeting.confirm(request.date(), request.time());
        return ConfirmedScheduleResponse.from(request.date(), request.time(), MeetingStatus.CONFIRMED, meeting.getTitle(), meeting.getDuration());
    }

    @Transactional
    public void cancelConfirm(Integer meetingId, Integer memberId) {
        Meeting meeting = findMeetingInternal(meetingId);

        if (!meeting.isHost(memberId)) {
            throw new BusinessException(ErrorCode.NOT_MEETING_HOST);
        }

        if (meeting.getStatus() != MeetingStatus.CONFIRMED) {
            throw new BusinessException(ErrorCode.NOT_CONFIRMED);
        }

        meeting.cancelConfirm();
    }

    @Transactional(readOnly = true)
    public ConfirmedScheduleResponse getConfirmedSchedule(Integer meetingId) {
        Meeting meeting = findMeetingInternal(meetingId);

        if (meeting.getStatus() != MeetingStatus.CONFIRMED || meeting.getConfirmedDate() == null) {
            throw new BusinessException(ErrorCode.NOT_CONFIRMED);
        }

        return ConfirmedScheduleResponse.from(
                meeting.getConfirmedDate(),
                meeting.getConfirmedTime(),
                meeting.getStatus(),
                meeting.getTitle(),
                meeting.getDuration()
        );
    }

    // ── 목록 조회 ──────────────────────────────
    @Transactional(readOnly = true)
    public List<MeetingEntryResponse> getMyMeetings(Integer memberId) {
        return meetingRepository.findByMember_IdOrderByCreatedAtDesc(memberId)
                .stream()
                .map(meeting -> {
                    List<LocalDate> dates = meeting.getMeetingsDates().stream()
                            .map(MeetingsDate::getDate)
                            .sorted()
                            .toList();
                    return new MeetingEntryResponse(
                            meeting.getId(),
                            meeting.getTitle(),
                            meeting.getCategory(),
                            meeting.getDuration(),
                            meeting.getStatus(),
                            meeting.getRandomUrl(),
                            dates,
                            meeting.getCreatedAt(),
                            meeting.getConfirmedDate(),
                            meeting.getConfirmedTime()
                    );
                })
                .toList();
    }

    // 외부 유틸
    @Transactional(readOnly = true)
    public Meeting getMeetingOrThrow(Integer meetingId) {
        return meetingRepository.findById(meetingId)
                .orElseThrow(() -> new BusinessException(ErrorCode.MEETING_NOT_FOUND));
    }

    @Transactional(readOnly = true)
    public Meeting getMeetingByRandomUrlOrThrow(String randomUrl) {
        return meetingRepository.findByRandomUrl(randomUrl)
                .orElseThrow(() -> new BusinessException(ErrorCode.MEETING_NOT_FOUND));
    }

    // ── 내부 유틸 ──────────────────────────────
    private String generateUniqueUrl() {
        String candidate = randomString(URL_LENGTH);
        while (meetingRepository.existsByRandomUrl(candidate)) {
            candidate = randomString(URL_LENGTH);
        }
        return candidate;
    }

    private String randomString(int length) {
        StringBuilder builder = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            int idx = secureRandom.nextInt(URL_CHAR_POOL.length());
            builder.append(URL_CHAR_POOL.charAt(idx));
        }
        return builder.toString();
    }

    private Meeting findMeetingInternal(Integer meetingId) {
        return meetingRepository.findById(meetingId)
                .orElseThrow(() -> new BusinessException(ErrorCode.MEETING_NOT_FOUND));
    }

    private Meeting findMeetingByRandomUrlInternal(String randomUrl) {
        return meetingRepository.findByRandomUrl(randomUrl)
                .orElseThrow(() -> new BusinessException(ErrorCode.MEETING_NOT_FOUND));
    }
}

File: src/main/java/com/back/nbe9112team06/domain/meeting/service/MeetingService.kt
package com.back.nbe9112team06.domain.meeting.service

import com.back.nbe9112team06.domain.meeting.dto.request.FinalizeRequest
import com.back.nbe9112team06.domain.meeting.dto.request.MeetingCreateRequest
import com.back.nbe9112team06.domain.meeting.dto.response.ConfirmedScheduleResponse
import com.back.nbe9112team06.domain.meeting.dto.response.MeetingCreateResponse
import com.back.nbe9112team06.domain.meeting.dto.response.MeetingEntryResponse
import com.back.nbe9112team06.domain.meeting.entity.Meeting
import com.back.nbe9112team06.domain.meeting.entity.MeetingStatus
import com.back.nbe9112team06.domain.meeting.entity.MeetingsDate
import com.back.nbe9112team06.domain.meeting.repository.MeetingRepository
import com.back.nbe9112team06.domain.member.service.MemberService
import com.back.nbe9112team06.domain.timetable.entity.TimeTable
import com.back.nbe9112team06.domain.timetable.service.TimeTableService
import com.back.nbe9112team06.global.error.ErrorCode
import com.back.nbe9112team06.global.exception.BusinessException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.security.SecureRandom

@Service
class MeetingService(
    private val meetingRepository: MeetingRepository,
    private val memberService: MemberService,
    private val timeTableService: TimeTableService
) {
    private val secureRandom = SecureRandom()

    // ── 모임 생성 ──────────────────────────────
    @Transactional
    fun createMeeting(memberId: Int, request: MeetingCreateRequest): MeetingCreateResponse {
        // TODO: Phase 2 - MemberService.findById()를 Optional 대신 Kotlin nullable 반환으로 변경하고 orElseThrow 제거
        val member = memberService.findById(memberId)
            .orElseThrow { BusinessException(ErrorCode.MEMBER_NOT_FOUND) }

        val randomUrl = generateUniqueUrl()
        val meeting = Meeting(
            request.title,
            request.category,
            request.duration,
            member,
            randomUrl
        )

        request.dates.forEach { date ->
            meeting.addMeetingsDate(MeetingsDate(date, member.email))
        }

        val saved = meetingRepository.save(meeting)

        val timeTable = TimeTable(meeting, mutableListOf())
        timeTableService.save(timeTable)

        return MeetingCreateResponse(saved.id, saved.randomUrl)
    }

    @Transactional(readOnly = true)
    fun getMeetingByRandomUrl(randomUrl: String): MeetingEntryResponse =
        findMeetingByRandomUrlInternal(randomUrl).toEntryResponse()

    // ── 모임 삭제 ──────────────────────────────
    @Transactional
    fun deleteMeeting(meetingId: Int, memberId: Int) {
        val meeting = findMeetingInternal(meetingId)
        meeting.validateHost(memberId)
        meetingRepository.delete(meeting)
    }

    @Transactional(readOnly = true)
    fun checkIsHost(randomUrl: String, memberId: Int): Boolean {
        val meeting = findMeetingByRandomUrlInternal(randomUrl)
        return meeting.isHost(memberId)
    }

    // ── 일정 확정 ──────────────────────────────
    @Transactional
    fun confirm(meetingId: Int, memberId: Int, request: FinalizeRequest): ConfirmedScheduleResponse {
        val meeting = findMeetingInternal(meetingId)
        meeting.validateHost(memberId)

        if (meeting.participants.isEmpty()) {
            throw BusinessException(ErrorCode.MEETING_NO_PARTICIPANTS)
        }

        if (meeting.status == MeetingStatus.CONFIRMED) {
            throw BusinessException(ErrorCode.ALREADY_CONFIRMED)
        }

        meeting.confirm(request.date, request.time)
        return ConfirmedScheduleResponse.from(
            request.date, request.time, MeetingStatus.CONFIRMED, meeting.title, meeting.duration
        )
    }

    @Transactional
    fun cancelConfirm(meetingId: Int, memberId: Int) {
        val meeting = findMeetingInternal(meetingId)
        meeting.validateHost(memberId)

        if (meeting.status != MeetingStatus.CONFIRMED) {
            throw BusinessException(ErrorCode.NOT_CONFIRMED)
        }

        meeting.cancelConfirm()
    }

    @Transactional(readOnly = true)
    fun getConfirmedSchedule(meetingId: Int): ConfirmedScheduleResponse {
        val meeting = findMeetingInternal(meetingId)

        if (meeting.status != MeetingStatus.CONFIRMED || meeting.confirmedDate == null) {
            throw BusinessException(ErrorCode.NOT_CONFIRMED)
        }

        return ConfirmedScheduleResponse.from(
            meeting.confirmedDate!!,
            meeting.confirmedTime!!,
            meeting.status,
            meeting.title,
            meeting.duration
        )
    }

    // ── 목록 조회 ──────────────────────────────
    @Transactional(readOnly = true)
    fun getMyMeetings(memberId: Int): List<MeetingEntryResponse> =
        meetingRepository.findByMember_IdOrderByCreatedAtDesc(memberId)
            .map { it.toEntryResponse() }

    // 외부 유틸
    @Transactional(readOnly = true)
    fun getMeetingOrThrow(meetingId: Int): Meeting =
        meetingRepository.findByIdOrNull(meetingId) ?: throw BusinessException(ErrorCode.MEETING_NOT_FOUND)

    @Transactional(readOnly = true)
    fun getMeetingByRandomUrlOrThrow(randomUrl: String): Meeting =
        meetingRepository.findByRandomUrl(randomUrl) ?: throw BusinessException(ErrorCode.MEETING_NOT_FOUND)

    // ── 내부 유틸 ──────────────────────────────
    private fun Meeting.validateHost(memberId: Int) {
        if (!isHost(memberId)) throw BusinessException(ErrorCode.NOT_MEETING_HOST)
    }

    private fun Meeting.toEntryResponse(): MeetingEntryResponse {
        val dates = meetingsDates.map { it.date }.sorted()
        return MeetingEntryResponse(
            id,
            title,
            category,
            duration,
            status,
            randomUrl,
            dates,
            createdAt,
            confirmedDate,
            confirmedTime
        )
    }

    private fun generateUniqueUrl(): String =
        generateSequence { randomString(URL_LENGTH) }
            .first { !meetingRepository.existsByRandomUrl(it) }

    private fun randomString(length: Int): String {
        val builder = StringBuilder(length)
        repeat(length) {
            val idx = secureRandom.nextInt(URL_CHAR_POOL.length)
            builder.append(URL_CHAR_POOL[idx])
        }
        return builder.toString()
    }

    private fun findMeetingInternal(meetingId: Int): Meeting =
        meetingRepository.findByIdOrNull(meetingId) ?: throw BusinessException(ErrorCode.MEETING_NOT_FOUND)

    private fun findMeetingByRandomUrlInternal(randomUrl: String): Meeting =
        meetingRepository.findByRandomUrl(randomUrl) ?: throw BusinessException(ErrorCode.MEETING_NOT_FOUND)

    companion object {
        private const val URL_CHAR_POOL = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        private const val URL_LENGTH = 10
    }
}
